naga/src/back/spv/ray.rs

```diff
@@ -441,151 +548,1312 @@ impl Writer {
+        if let Some(&word) = self
+            .ray_query_functions
+            .get(&LookupRayQueryFunction::GenerateIntersection)
+        {
+            return word;
+        }
+
+        let ray_query_type_id = self.get_ray_query_pointer_id();
+
+        let u32_ty = self.get_u32_type_id();
+        let u32_ptr_ty = self.get_pointer_type_id(u32_ty, spirv::StorageClass::Function);
+
+        let f32_type_id = self.get_f32_type_id();
+
+        let bool_type_id = self.get_bool_type_id();
+
+        let func_ty = self.get_function_type(LookupFunctionType {
+            parameter_type_ids: vec![ray_query_type_id, u32_ptr_ty, f32_type_id],
+            return_type_id: self.void_type,
+        });
+
+        let mut function = Function::default();
+        let func_id = self.id_gen.next();
+        function.signature = Some(Instruction::function(
+            self.void_type,
+            func_id,
+            spirv::FunctionControl::empty(),
+            func_ty,
+        ));
+
+        let query_id = self.id_gen.next();
+        let instruction = Instruction::function_parameter(ray_query_type_id, query_id);
+        function.parameters.push(FunctionArgument {
+            instruction,
+            handle_id: 0,
+        });
+
+        let init_tracker_id = self.id_gen.next();
+        let instruction = Instruction::function_parameter(u32_ptr_ty, init_tracker_id);
+        function.parameters.push(FunctionArgument {
+            instruction,
+            handle_id: 1,
+        });
+
+        let depth_id = self.id_gen.next();
+        let instruction = Instruction::function_parameter(f32_type_id, depth_id);
+        function.parameters.push(FunctionArgument {
+            instruction,
+            handle_id: 2,
+        });
+
+        let block_id = self.id_gen.next();
+        let mut block = Block::new(block_id);
+
+        let valid_id = self.id_gen.next();
+        let mut valid_block = Block::new(valid_id);
+
+        let final_label_id = self.id_gen.next();
+        let final_block = Block::new(final_label_id);
+
+        let instruction = if self.ray_query_initialization_tracking {
+            let initialized_tracker_id = self.id_gen.next();
+            block.body.push(Instruction::load(
+                u32_ty,
+                initialized_tracker_id,
+                init_tracker_id,
+                None,
+            ));
+
+            let proceeded_id = write_ray_flags_contains_flags(
+                self,
+                &mut block,
+                initialized_tracker_id,
+                super::RayQueryPoint::PROCEED.bits(),
+            );
+            let finished_proceed_id = write_ray_flags_contains_flags(
+                self,
+                &mut block,
+                initialized_tracker_id,
+                super::RayQueryPoint::FINISHED_TRAVERSAL.bits(),
+            );
+            // TODO: Is double calling this invalid? Can't find anything to suggest so.
```

Comment:
```suggestion
            // Can't find anything to suggest that calling this twice is invalid.
```

Also what is "this" finished traversal?

Reply:
Sorry, I should have been more clear on this. "this" is the function (or spir-v OP technically)